File: server/aoNewsletter2/Views/NewsletterWidget.vb

Imports Contensive.Addons.Newsletter.Models.Db
Imports Contensive.Addons.Newsletter.Controllers
Imports Contensive.BaseClasses
Imports Contensive.Addons.Newsletter.Models.View

Namespace Views
    '
    '====================================================================================================
    ''' <summary>
    ''' Design block with a centered headline, image, paragraph text and a button.
    ''' </summary>
    Public Class NewsletterWidget
        Inherits AddonBaseClass
        '
        '====================================================================================================
        '
        Public Overrides Function Execute(ByVal CP As CPBaseClass) As Object
            Const designBlockName As String = "Newsletter Widget"
            Try
                '
                ' -- read instanceId, guid created uniquely for this instance of the addon on a page
                Dim result = String.Empty
                Dim settingsGuid = DesignBlockController.getSettingsGuid(CP, designBlockName, result)
                If (String.IsNullOrEmpty(settingsGuid)) Then Return result
                '
                ' -- locate or create a data record for this guid
                Dim newsletterId As Integer = NewsletterController.getNewsletterId(CP, settingsGuid)
                Dim currentIssueID As Integer = NewsletterController.GetCurrentIssueID(CP, newsletterId)
                Dim settings = NewsletterModel.createOrAddSettings(CP, settingsGuid)
                If (settings Is Nothing) Then Throw New ApplicationException("Could not create the design block settings record.")
                '
                ' -- create legacy newsletter
                '
                Dim legacyNewsletter As String = (New NewsletterClass()).getLegacyNewsletter(CP, newsletterId, currentIssueID)
                '
                ' -- translate the Db model to a view model and mustache it into the layout
                Dim viewModel = NewsletterViewModel.create(CP, settings, legacyNewsletter)
                If (viewModel Is Nothing) Then Throw New ApplicationException("Could not create design block view model.")
                result = Nustache.Core.Render.StringToString(My.Resources.NewsletterLayout, viewModel)
                '
                ' -- if editing enabled, add the link and wrapperwrapper
                Return CP.Content.GetEditWrapper(result, NewsletterModel.tableMetadata.contentName, settings.id)
            Catch ex As Exception
                CP.Site.ErrorReport(ex)
                Return "<!-- " & designBlockName & ", Unexpected Exception -->"
            End Try
        End Function
    End Class
End Namespace

File: server/aoNewsletter2/Views/NewsletterWidget.cs
using System;
using Contensive.Addons.Newsletter.Controllers;

using Contensive.Addons.Newsletter.Models.Db;
using Contensive.Addons.Newsletter.Models.View;
using Contensive.BaseClasses;

namespace Contensive.Addons.Newsletter.Views {
    // 
    // ====================================================================================================
    /// <summary>
    /// Design block with a centered headline, image, paragraph text and a button.
    /// </summary>
    public class NewsletterWidget : AddonBaseClass {
        // 
        // ====================================================================================================
        // 
        public override object Execute(CPBaseClass CP) {
            const string designBlockName = "Newsletter Widget";
            try {
                // 
                // -- read instanceId, guid created uniquely for this instance of the addon on a page
                string result = string.Empty;
                string settingsGuid = DesignBlockController.getSettingsGuid(CP, designBlockName, ref result);
                if (string.IsNullOrEmpty(settingsGuid))
                    return result;
                // 
                // -- locate or create a data record for this guid
                int newsletterId = NewsletterController.getNewsletterId(CP, settingsGuid);
                int currentIssueID = NewsletterController.GetCurrentIssueID(CP, newsletterId);
                var settings = NewsletterModel.createOrAddSettings(CP, settingsGuid);
                if (settings is null)
                    throw new ApplicationException("Could not create the design block settings record.");
                // 
                // -- create legacy newsletter
                // 
                string legacyNewsletter = new NewsletterClass().getLegacyNewsletter(CP, newsletterId, currentIssueID)?.ToString() ?? "";
                // 
                // -- translate the Db model to a view model and mustache it into the layout
                var viewModel = NewsletterViewModel.create(CP, settings, legacyNewsletter);
                if (viewModel is null)
                    throw new ApplicationException("Could not create design block view model.");
                result = Nustache.Core.Render.StringToString(My.Resources.Resources.NewsletterLayout, viewModel);
                // 
                // -- if editing enabled, add the link and wrapperwrapper
                return CP.Content.GetEditWrapper(result, NewsletterModel.tableMetadata.contentName, settings.id);
            } catch (Exception ex) {
                CP.Site.ErrorReport(ex);
                return "<!-- " + designBlockName + ", Unexpected Exception -->";
            }
        }
    }
}